Fix resending configs upon MQTT reconnects

QRPp opened the pull request:
All objects suppressed resending their configuration, having sent it
successfully via an earlier connection.

pimvanpelt commented:
Thanks for the PR. I don't understand why it is needed,
because mgos_homeassistant_api.c#L596 sets 'retain' flag on the initial
MQTT publication of the configs.
Can you explain to me how, upon a restart/reconnect of the microcontroller,
your pubsub server lost its messages?

On Fri, Apr 2, 2021 at 3:06 AM Awesome Walrus ***@***.***>
wrote:

> All objects suppressed resending their configuration, having sent it
> successfully via an earlier connection.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/mongoose-os-libs/homeassistant/pull/2
> Commit Summary
>
>    - Fix resending configs upon MQTT reconnects
>
> File Changes
>
>    - *M* include/mgos_homeassistant_api.h
>    <https://github.com/mongoose-os-libs/homeassistant/pull/2/files#diff-13e60350d88e85c56af825738396c0752ab193ed96209e08e60d447ae84cf196>
>    (2)
>    - *M* src/mgos_homeassistant.c
>    <https://github.com/mongoose-os-libs/homeassistant/pull/2/files#diff-08cf5d3a85b6f01760fda8b7c5f2de18d06cf460918cd5b4d1e4b2be4ed02b9f>
>    (4)
>    - *M* src/mgos_homeassistant_api.c
>    <https://github.com/mongoose-os-libs/homeassistant/pull/2/files#diff-0b98a0e1a3c92a8e79c14f347e893c18c9c9e0ff5a647d5ca286767a232ee8e2>
>    (3)
>
> Patch Links:
>
>    - https://github.com/mongoose-os-libs/homeassistant/pull/2.patch
>    - https://github.com/mongoose-os-libs/homeassistant/pull/2.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/mongoose-os-libs/homeassistant/pull/2>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACDRSPL56ZE4LSLDUJXKIADTGUKB3ANCNFSM42IBO4BA>
> .
>


-- 
Pim van Pelt ***@***.***>
PBVP1-RIPE - http://www.ipng.nl/

QRPp commented:
Thanks for the quick reply!

This fix isn't applicable to restarts of a microcontroller, only reconnects trigger the codepath at hand.  One reason for reconnects to happen is if, e.g., an MQTT broker itself restarts.

The “retain” flag may not always help across broker restarts for various reasons.  To name a few:
- a broker may not be set up to save and reload its state, or even capable thereof, having no writable storage;
- a broker may be in a dubious redundant setup where instances don't exchange state;
- a broker may have been reinstalled or otherwise experienced data loss (my initial encounters of the problem were of this kind);
- a broker may have been restored to a different state, e.g., one backed up earlier on, when the relevant configuration differed (had this kind too).

Since Mongoose OS HA library _already attempts_ to [resend the configuration](https://github.com/mongoose-os-libs/homeassistant/pull/2/files#diff-08cf5d3a85b6f01760fda8b7c5f2de18d06cf460918cd5b4d1e4b2be4ed02b9fL42) upon _any reconnect,_ handling such situations seemed a part of the intent, TBH.  Given the `.config_sent` logic, however, these attempts almost never lead to any practical effect on the broker (that doesn't apply to resending the status, of course).

QRPp commented:
In short, resending 'retain'-ed configuration messages to the broker upon reconnecting to it ensures that the broker, which may have lost or changed the state meanwhile, has its state corrected if need be.  From a certain point of view, the purpose and the result are more or less the same as those of resending the same 'retain'-ed configuration messages by the broker itself to its clients upon their reconnects to the broker: they also may have lost the corresponding state meanwhile.

Though loss of the state by the broker may appear less likely than loss of the state by the broker clients, it's not impossible, and ensuring prompt propagation of the state from the source of truth (in this case, the MCU running an instance of Mongoose OS homeassistant library) protects against “split brain” situations.

QRPp committed:
Fix resending configs upon MQTT reconnects

All objects suppressed resending their configuration, having sent it
successfully via an earlier connection.
